File: .githooks/github/pull_requests.py
#!/usr/bin/env python3
"""Validate PR content against .githooks/spec/github_pull_requests.yaml.

Replaces bin/validate_pr.sh (non-review P-* rules).
"""

from __future__ import annotations



import sys as _sys
_sys.dont_write_bytecode = True  # 不生成 __pycache__
import re
import sys
from pathlib import Path
from typing import Any, Optional

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))
sys.path.insert(0, str(ROOT / ".."))

from _shared import (  # noqa: E402
    Finding,
    Severity,
    aggregate_result,
    gh_api_get,
    load_yaml,
    print_findings,
)

CUTOFF = 55

# ---------------------------------------------------------------------------
# helpers
# ---------------------------------------------------------------------------

CJK_RE = re.compile(r"[\u4e00-\u9fff]")
FULLWIDTH_BRACKETS = "（）「」【】『』《》〈〉《》﹁﹂"
H1_RE = re.compile(r"^# [^#]", re.MULTILINE)
HEADING_RE = re.compile(r"^#{1,6} ", re.MULTILINE)
CHECKBOX_RE = re.compile(r"^\s*-\s*\[([ xX])\]", re.MULTILINE)
TABLE_RE = re.compile(r"^\|[- ]+\|", re.MULTILINE)

TICKED = ("x", "X")


def _has_cjk(s: str) -> bool:
    return bool(CJK_RE.search(s))


def _section(body: str, heading: str) -> str:
    m = re.search(rf"^## {re.escape(heading)}\s*$", body, re.MULTILINE)
    if not m:
        return ""
    rest = body[m.end():]
    nxt = re.search(r"^## ", rest, re.MULTILINE)
    return rest[: nxt.start()] if nxt else rest


def _headings(body: str) -> list[str]:
    return [
        line.strip().lstrip("#").strip()
        for line in body.splitlines()
        if HEADING_RE.match(line)
    ]


def _explicit_bool(v: Any, default: bool) -> bool:
    if isinstance(v, bool):
        return v
    if isinstance(v, str):
        return v.strip().lower() in ("true", "1", "yes")
    return default


# ---------------------------------------------------------------------------
# rule registry
# ---------------------------------------------------------------------------

def _load_config() -> dict[str, Any]:
    cfg = load_yaml(ROOT / "spec" / "github_pull_requests.yaml")
    cfg.setdefault("required_body_headings", ["Issue", "What", "Why"])
    cfg.setdefault("required_title_headings", ["Goal", "Background"])
    cfg.setdefault("forbidden_brackets_in_title", list(FULLWIDTH_BRACKETS))
    cfg.setdefault("forbidden_keywords", [])
    cfg.setdefault("ci_check_mode", "WARN")
    cfg.setdefault("done_when_check_mode", "FAIL")
    cfg.setdefault("keyword_label_suggestions", {})
    return cfg


def _extract_fixes(body: str) -> list[str]:
    """Extract all 'Fixes #N' / 'Closes #N' / 'Resolves #N' lines."""
    return re.findall(r"(?:Fixes|Closes|Resolves)\s+#(\d+)", body)


# ---------------------------------------------------------------------------
# rule functions
# ---------------------------------------------------------------------------

def run(repo: str, num: int, mode: str = "", strict: bool = False) -> list[Finding]:
    cfg = _load_config()
    findings: list[Finding] = []

    if not strict and num < CUTOFF:
        print(f"== PR #{num}: SKIP (below cutoff {CUTOFF}; legacy exempt.) ==")
        return []

    pr = gh_api_get(f"repos/{repo}/pulls/{num}")
    if not pr:
        findings.append(Finding("P-00", Severity.FAIL, "PR fetch returned nothing"))
        return findings

    title: str = pr.get("title", "")
    body: str = pr.get("body") or ""
    state: str = pr.get("state", "open").lower()
    labels: list[str] = [l.get("name", "") for l in pr.get("labels", []) or []]
    mergeable: bool = pr.get("mergeable", True)

    print(f"== PR #{num}: {title} (state={state}, mergeable={mergeable}) ==")

    # ---- P-01 / P-02 title ----
    print("--- title ---")
    # P-01: title must be English (no CJK)
    if _has_cjk(title):
        findings.append(Finding("P-01", Severity.FAIL, "title contains CJK (title should be English)"))
    else:
        findings.append(Finding("P-01", Severity.INFO, "title is English"))

    # P-02: conventional commit check (WARN only — repo allows natural English)
    if re.match(r"^(feat|fix|chore|docs|style|refactor|test|ci|build|perf|revert)(\(.+\))?:\s+", title):
        findings.append(Finding("P-02", Severity.INFO, "conventional commit title"))
    else:
        findings.append(Finding("P-02", Severity.WARN, f"title not conventional commit (repo template allows natural English): {title}"))

    # ---- Body structure headings ----
    print("--- body structure ---")
    body_headings = set(_headings(body))
    for h in cfg.get("required_body_headings", ["What", "Why", "Issue", "Construction plan", "Delivery record", "How to test", "Checklist"]):
        if h in body_headings:
            findings.append(Finding("P-xx", Severity.INFO, f"heading present: {h}"))
        else:
            findings.append(Finding("P-xx", Severity.FAIL, f"missing heading: ## {h}"))

    # ---- P-10 heading English / What Chinese ----
    print("--- headings ---")
    bad_h = [h for h in _headings(body) if _has_cjk(h)]
    if bad_h:
        findings.append(Finding("P-10", Severity.FAIL, f"headings contain CJK (headings must be English): {bad_h}"))
    else:
        findings.append(Finding("P-10", Severity.INFO, "headings are English only"))

    what = _section(body, "What")
    if _has_cjk(what):
        findings.append(Finding("P-10", Severity.INFO, "What section has Chinese prose"))
    else:
        findings.append(Finding("P-10", Severity.WARN, "What section has no Chinese prose (template requires Chinese)"))

    # ---- P-14 / P-14b labels ----
    print("--- labels ---")
    all_labels = gh_api_get(f"repos/{repo}/labels")
    valid_names = {l.get("name", "") for l in all_labels or []}
    unknown = [l for l in labels if l and l not in valid_names]
    if unknown:
        findings.append(Finding("P-14", Severity.FAIL, f"labels not in repo: {unknown}"))
    else:
        findings.append(Finding("P-14", Severity.INFO, "labels all exist in repo"))
    type_labels_cfg = cfg.get("type_labels_cfg", ["bug", "enhancement", "feature", "documentation", "chore", "refactor", "tests", "epic"])
    if any(l in type_labels_cfg for l in labels):
        findings.append(Finding("P-14b", Severity.INFO, "type label present"))
    else:
        findings.append(Finding("P-14b", Severity.FAIL, "no type label (expected one of the type set)"))

    # P-14b keyword suggestions
    kw_map = cfg.get("keyword_label_suggestions", {})
    if kw_map:
        haystack = f"{title}\n{body}".lower()
        missing_suggestions: list[str] = []
        for keyword, suggested_label in kw_map.items():
            if keyword.lower() in haystack and suggested_label not in labels:
                missing_suggestions.append(suggested_label)
        if missing_suggestions:
            findings.append(Finding("P-14b", Severity.WARN, f"based on content keywords, consider also labeling: {' '.join(sorted(set(missing_suggestions)))}"))
        else:
            findings.append(Finding("P-14b", Severity.INFO, "content keywords align with assigned labels"))
    else:
        findings.append(Finding("P-14b", Severity.INFO, "no keyword suggestions (or policy not configured)"))

    # ---- P-11 / P-12 / P-13 Issue linkage ----
    print("--- issue link ---")
    fixes = _extract_fixes(body)
    fixes_unique = sorted(set(fixes), key=int)
    fixes_count = len(fixes_unique)

    # P-11: premature Fixes while open
    if state == "open" and fixes_count > 0:
        findings.append(Finding("P-11", Severity.WARN, "open PR already uses Fixes # (may close issue prematurely)"))
    else:
        findings.append(Finding("P-11", Severity.INFO, "no premature Fixes while open (or PR not open)"))

    # P-12: exactly one Fixes #
    draft = pr.get("draft", False)
    if fixes_count == 1:
        findings.append(Finding("P-12", Severity.INFO, f"exactly one Fixes #"))
    elif fixes_count == 0:
        if draft:
            findings.append(Finding("P-12", Severity.INFO, "draft PR, Fixes may appear at merge authorization"))
        else:
            findings.append(Finding("P-12", Severity.WARN, "no Fixes # yet (needs one primary issue before merge)"))
    else:
        findings.append(Finding("P-12", Severity.WARN, f"multiple Fixes # ({fixes_count}): one PR should close one issue"))

    # P-13: one primary issue
    if fixes_count <= 1:
        findings.append(Finding("P-13", Severity.INFO, "one primary issue"))
    else:
        findings.append(Finding("P-13", Severity.FAIL, "one PR should close one primary issue"))

    # ---- P-16 / P-17 / P-18 / P-19 repo repo not here → skipped
    # ---- P-19 whitespace ratio — skipped (needs git diff)
    # ---- P-39 closing reference ----
    print("--- closing reference ---")
    # P-39: check closing reference
    # Find the primary issue from Fixes #
    if fixes:
        primary = fixes[0]
        # closing reference check beyond script scope
        # simplified: just note presence of Fixes
        findings.append(Finding("P-39", Severity.INFO, f"Fixes # present: #{primary}; closing reference check requires base branch = default)"))

    # ---- P-16 repo not a git repo → skipped (handled in hook) ----

    # ---- P-17 / P-18 / P-19 repo not a git repo → skipped ----

    # ---- P-19 whitespace ratio — skipped ----

    # ---- P-20 / P-21 / P-21b labels ----
    print("--- labels (PR) ---")
    valid_names = {l.get("name", "") for l in (gh_api_get(f"repos/{repo}/labels") or [])}
    unknown = [l for l in labels if l and l not in valid_names]
    if unknown:
        findings.append(Finding("P-20", Severity.FAIL, f"labels not in repo: {unknown}"))
    else:
        findings.append(Finding("P-20", Severity.INFO, "labels all exist in repo"))
    type_labels_cfg = cfg.get("type_labels_cfg", ["bug", "enhancement", "feature", "documentation", "chore", "refactor", "tests", "epic"])
    if any(l in type_labels_cfg for l in labels):
        findings.append(Finding("P-21", Severity.INFO, "type label present"))
    else:
        findings.append(Finding("P-21", Severity.FAIL, "no type label (expected one of the type set)"))

    # P-21b keyword suggestions (same as issue)
    kw_map = cfg.get("keyword_label_suggestions", {})
    if kw_map:
        haystack = f"{title}\n{body}".lower()
        missing_suggestions: list[str] = []
        for keyword, suggested_label in kw_map.items():
            if keyword.lower() in haystack and suggested_label not in labels:
                missing_suggestions.append(suggested_label)
        if missing_suggestions:
            findings.append(Finding("P-21b", Severity.WARN, f"based on content keywords, consider also labeling: {' '.join(sorted(set(missing_suggestions)))}"))
        else:
            findings.append(Finding("P-21b", Severity.INFO, "content keywords align with assigned labels"))
    else:
        findings.append(Finding("P-21b", Severity.INFO, "no keyword suggestions (or policy not configured)"))

    # ---- P-25 / P-24 / P-22 inline review / reply — review 相关跳过，见 #103 ----
    # ---- P-26 / P-27 CI 状态 — 此处仅 WARN，实际检查由 CI 系统 —
    # ---- P-30 Done when — PR 没有 Done whenアイデ (只针对 sub-issue, PR 模式跳过) ----

    # ---- P-31 branch name ----
    print("--- branch ---")
    head_ref = pr.get("head", {}).get("ref", "")
    allowed = cfg.get("allowed_branch_prefixes", ["feat/", "fix/", "chore/", "epic/", "main", "master", "release/"])
    # 检查分支名是否以允许的前缀开头
    import fnmatch
    prefix_ok = any(fnmatch.fnmatch(head_ref, f"*{p}*") for p in allowed)
    # 分支名必须以 allowed 前缀开头
    if not head_ref or not any(head_ref.startswith(p) for p in allowed):
        findings.append(Finding("P-31", Severity.FAIL, f"branch name not allowed: {head_ref} (allowed prefixes: {allowed})"))
    else:
        findings.append(Finding("P-31", Severity.INFO, f"branch name OK: {head_ref} (prefixes: {allowed})"))

    # ---- P-35 / P-36 review 前缀 — review 相关跳过，见 #103 ----

    # ---- P-38 maintainer review — 由人工决定, 脚本只 WARN ----
    findings.append(Finding("P-38", Severity.WARN, "no maintainer review (COMMENTED/APPROVED/CHANGES_REQUESTED) — human required"))

    # ---- aggregate result — tasks depend on repo hooks interpreting exit code ----
    return findings


def main() -> int:
    args = [a for a in sys.argv[1:] if not a.startswith("--")]
    strict = "--strict" in sys.argv
    if len(args) < 2:
        print(__doc__)
        return 2
    repo = args[0]
    num = int(args[1])
    mode = args[2] if len(args) > 2 else ""
    findings = run(repo, num, mode, strict=strict)
    print_findings(findings)
    return aggregate_result(findings)


if __name__ == "__main__":
    sys.exit(main())

File: crates/evidence/spec/src/rules/pull_requests.rs
//! PR-* validation rules.
//!
//! Pure content checks driven by `spec/github_pull_requests.yaml`, including
//! the fork "user:" prefix strip on `head_ref` (PR-08).

use regex::Regex;
use serde_yaml::Value as YamlValue;
use std::collections::HashSet;

use crate::shared::{Finding, Severity};

// ---------------------------------------------------------------------------
// Default config values (mirror the YAML so the function works without a file)
// ---------------------------------------------------------------------------

const DEFAULT_BODY_HEADINGS: &[&str] = &[
    "Issue",
    "What",
    "Why",
    "Construction plan",
    "Delivery record",
    "How to test",
    "Checklist",
];

const DEFAULT_BRANCH_PREFIXES: &[&str] = &[
    "feat/", "fix/", "chore/", "epic/", "main", "master", "release/",
];

const DEFAULT_TYPE_LABELS: &[&str] = &[
    "bug",
    "enhancement",
    "feature",
    "documentation",
    "chore",
    "refactor",
    "tests",
    "epic",
];

// ponytail: keyword map as a flat array of (keyword, label) pairs; avoids
// allocating a HashMap when the config is absent or empty.
const DEFAULT_KEYWORD_SUGGESTIONS: &[(&str, &str)] = &[
    ("bug", "bug"),
    ("修复", "bug"),
    ("新功能", "enhancement"),
    ("功能", "enhancement"),
    ("功能添加", "enhancement"),
    ("清理", "chore"),
    ("优化", "refactor"),
    ("测试", "tests"),
    ("代码质量", "refactor"),
];

// ---------------------------------------------------------------------------
// Regexes — compiled once via std::sync::LazyLock (no once_cell dependency).
// ---------------------------------------------------------------------------

fn cjk_re() -> &'static Regex {
    static RE: std::sync::LazyLock<Regex> =
        std::sync::LazyLock::new(|| Regex::new(r"[\u4e00-\u9fff]").unwrap());
    &RE
}

fn heading_re() -> &'static Regex {
    static RE: std::sync::LazyLock<Regex> =
        std::sync::LazyLock::new(|| Regex::new(r"(?m)^#{1,6} ").unwrap());
    &RE
}

fn checkbox_re() -> &'static Regex {
    static RE: std::sync::LazyLock<Regex> =
        std::sync::LazyLock::new(|| Regex::new(r"(?m)^\s*-\s*\[([ xX])\]").unwrap());
    &RE
}

fn conv_commit_re() -> &'static Regex {
    static RE: std::sync::LazyLock<Regex> = std::sync::LazyLock::new(|| {
        Regex::new(r"^(feat|fix|chore|docs|style|refactor|test|ci|build|perf|revert)(\(.+\))?:\s+")
            .unwrap()
    });
    &RE
}

fn fixes_re() -> &'static Regex {
    static RE: std::sync::LazyLock<Regex> =
        std::sync::LazyLock::new(|| Regex::new(r"(?:Fixes|Closes|Resolves)\s+#(\d+)").unwrap());
    &RE
}

fn text_links_re() -> &'static Regex {
    static RE: std::sync::LazyLock<Regex> =
        std::sync::LazyLock::new(|| Regex::new(r"(?:Part of|Related)\s+#(\d+)").unwrap());
    &RE
}

// ---------------------------------------------------------------------------
// Helpers — direct ports of the Python _functions
// ---------------------------------------------------------------------------

/// True if `s` contains any CJK ideograph (U+4E00–U+9FFF).
fn has_cjk(s: &str) -> bool {
    cjk_re().is_match(s)
}

/// Extract the body text under a `## heading` line, up to the next `## `.
/// Mirrors Python `_section`.
fn section(body: &str, heading: &str) -> String {
    let pattern = format!(r"(?m)^## {}\s*$", regex::escape(heading));
    let re = match Regex::new(&pattern) {
        Ok(r) => r,
        Err(_) => return String::new(),
    };
    let m = match re.find(body) {
        Some(m) => m,
        None => return String::new(),
    };
    let rest = &body[m.end()..];
    let next_re = match Regex::new(r"(?m)^## ") {
        Ok(r) => r,
        Err(_) => return rest.to_string(),
    };
    match next_re.find(rest) {
        Some(n) => rest[..n.start()].to_string(),
        None => rest.to_string(),
    }
}

/// Return all heading texts (leading `#`s stripped) in line order.
/// Mirrors Python `_headings`.
fn headings(body: &str) -> Vec<String> {
    let re = heading_re();
    body.lines()
        .filter(|line| re.is_match(line))
        .map(|line| line.trim().trim_start_matches('#').trim().to_string())
        .collect()
}

/// Extract all `Fixes #N` / `Closes #N` / `Resolves #N` issue numbers.
fn extract_fixes(body: &str) -> Vec<String> {
    fixes_re()
        .captures_iter(body)
        .map(|c| c.get(1).unwrap().as_str().to_string())
        .collect()
}

// ---------------------------------------------------------------------------
// Config extraction helpers
// ---------------------------------------------------------------------------

fn cfg_body_headings(cfg: Option<&YamlValue>) -> Vec<String> {
    if let Some(arr) = cfg
        .and_then(|c| c.get("required_body_headings"))
        .and_then(|v| v.as_sequence())
    {
        let vals: Vec<String> = arr
            .iter()
            .filter_map(|v| v.as_str().map(String::from))
            .collect();
        if !vals.is_empty() {
            return vals;
        }
    }
    DEFAULT_BODY_HEADINGS
        .iter()
        .map(|s| s.to_string())
        .collect()
}

fn cfg_branch_prefixes(cfg: Option<&YamlValue>) -> Vec<String> {
    if let Some(arr) = cfg
        .and_then(|c| c.get("allowed_branch_prefixes"))
        .and_then(|v| v.as_sequence())
    {
        let vals: Vec<String> = arr
            .iter()
            .filter_map(|v| v.as_str().map(String::from))
            .collect();
        if !vals.is_empty() {
            return vals;
        }
    }
    DEFAULT_BRANCH_PREFIXES
        .iter()
        .map(|s| s.to_string())
        .collect()
}

fn cfg_type_labels(cfg: Option<&YamlValue>) -> Vec<String> {
    if let Some(arr) = cfg
        .and_then(|c| c.get("type_labels_cfg"))
        .and_then(|v| v.as_sequence())
    {
        let vals: Vec<String> = arr
            .iter()
            .filter_map(|v| v.as_str().map(String::from))
            .collect();
        if !vals.is_empty() {
            return vals;
        }
    }
    DEFAULT_TYPE_LABELS.iter().map(|s| s.to_string()).collect()
}

/// Extract keyword→label suggestions map from config YAML.
/// Falls back to `DEFAULT_KEYWORD_SUGGESTIONS` when absent or empty.
fn cfg_keyword_suggestions(cfg: Option<&YamlValue>) -> Vec<(String, String)> {
    if let Some(map) = cfg
        .and_then(|c| c.get("keyword_label_suggestions"))
        .and_then(|v| v.as_mapping())
    {
        let pairs: Vec<(String, String)> = map
            .iter()
            .filter_map(|(k, v)| {
                let key = k.as_str()?.to_string();
                let val = v.as_str()?.to_string();
                Some((key, val))
            })
            .collect();
        if !pairs.is_empty() {
            return pairs;
        }
    }
    DEFAULT_KEYWORD_SUGGESTIONS
        .iter()
        .map(|(k, v)| (k.to_string(), v.to_string()))
        .collect()
}

// ---------------------------------------------------------------------------
// Main entry — the `check_content` port
// ---------------------------------------------------------------------------

/// Pure content validation for a pull request.  No API calls.
///
/// * `title`     — PR title
/// * `body`      — PR body (markdown)
/// * `labels`    — label names on the PR
/// * `head_ref`  — head ref name (may include fork "user:" prefix)
/// * `state`     — "open" or "closed"/"merged"
/// * `draft`     — whether the PR is a draft
/// * `cfg`       — optional parsed YAML config (the spec file)
///
/// Returns a `Vec<Finding>` in the same order as the Python version.
pub fn check_content(
    title: &str,
    body: &str,
    labels: &[&str],
    head_ref: &str,
    state: &str,
    draft: bool,
    cfg: Option<&YamlValue>,
) -> Vec<Finding> {
    let mut findings: Vec<Finding> = Vec::new();

    // PR-01 title English / PR-02 conventional commit
    if has_cjk(title) {
        findings.push(Finding::new(
            "PR-01",
            Severity::Fail,
            "title contains CJK (title should be English)",
        ));
    } else {
        findings.push(Finding::new("PR-01", Severity::Info, "title is English"));
    }
    if conv_commit_re().is_match(title) {
        findings.push(Finding::new(
            "PR-02",
            Severity::Info,
            "conventional commit title",
        ));
    } else {
        findings.push(Finding::new(
            "PR-02",
            Severity::Warn,
            &format!(
                "title not conventional commit (repo template allows natural English): {title}"
            ),
        ));
    }

    // PR-03 body structure headings
    let body_h: HashSet<String> = headings(body).into_iter().collect();
    let required = cfg_body_headings(cfg);
    for h in &required {
        if body_h.contains(h) {
            findings.push(Finding::new(
                "PR-03",
                Severity::Info,
                &format!("heading present: {h}"),
            ));
        } else {
            findings.push(Finding::new(
                "PR-03",
                Severity::Fail,
                &format!("missing heading: ## {h}"),
            ));
        }
    }

    // PR-07 Construction plan / Checklist ≥2 checkboxes
    let cb_re = checkbox_re();
    if body_h.contains("Construction plan") {
        let plan = section(body, "Construction plan");
        let boxes = cb_re.captures_iter(&plan).count();
        if boxes < 2 {
            findings.push(Finding::new(
                "PR-07",
                Severity::Fail,
                &format!("Construction plan 必须至少 2 个 checkbox，当前 {boxes} 个"),
            ));
        }
    }
    if body_h.contains("Checklist") {
        let checklist = section(body, "Checklist");
        let boxes = cb_re.captures_iter(&checklist).count();
        if boxes < 2 {
            findings.push(Finding::new(
                "PR-07",
                Severity::Fail,
                &format!("Checklist 必须至少 2 个 checkbox，当前 {boxes} 个"),
            ));
        }
    }

    // PR-04 headings English only + What section Chinese prose
    let all_headings = headings(body);
    let bad_h: Vec<&String> = all_headings.iter().filter(|h| has_cjk(h)).collect();
    if !bad_h.is_empty() {
        let joined: Vec<String> = bad_h.iter().map(|h| h.to_string()).collect();
        findings.push(Finding::new(
            "PR-04",
            Severity::Fail,
            &format!(
                "headings contain CJK (headings must be English): [{}]",
                joined.join(", ")
            ),
        ));
    } else {
        findings.push(Finding::new(
            "PR-04",
            Severity::Info,
            "headings are English only",
        ));
    }
    let what = section(body, "What");
    if has_cjk(&what) {
        findings.push(Finding::new(
            "PR-04",
            Severity::Info,
            "What section has Chinese prose",
        ));
    } else {
        findings.push(Finding::new(
            "PR-04",
            Severity::Warn,
            "What section has no Chinese prose (template requires Chinese)",
        ));
    }

    // PR-05 issue linkage — Fixes #N checks
    let fixes = extract_fixes(body);
    // dedupe + sort numerically (Python: sorted(set(fixes), key=int))
    let mut fixes_unique: Vec<i32> = fixes
        .iter()
        .filter_map(|s| s.parse::<i32>().ok())
        .collect::<HashSet<_>>()
        .into_iter()
        .collect();
    fixes_unique.sort();
    let fixes_count = fixes_unique.len();

    if state == "open" && fixes_count > 0 {
        findings.push(Finding::new(
            "PR-05",
            Severity::Warn,
            "open PR already uses Fixes # (may close issue prematurely)",
        ));
    } else {
        findings.push(Finding::new(
            "PR-05",
            Severity::Info,
            "no premature Fixes while open (or PR not open)",
        ));
    }
    if fixes_count == 1 {
        findings.push(Finding::new("PR-05", Severity::Info, "exactly one Fixes #"));
    } else if fixes_count == 0 {
        if draft {
            findings.push(Finding::new(
                "PR-05",
                Severity::Info,
                "draft PR, Fixes may appear at merge authorization",
            ));
        } else {
            findings.push(Finding::new(
                "PR-05",
                Severity::Warn,
                "no Fixes # yet (needs one primary issue before merge)",
            ));
        }
    } else {
        findings.push(Finding::new(
            "PR-05",
            Severity::Warn,
            &format!("multiple Fixes # ({fixes_count}): one PR should close one issue"),
        ));
    }
    if fixes_count <= 1 {
        findings.push(Finding::new("PR-05", Severity::Info, "one primary issue"));
    } else {
        findings.push(Finding::new(
            "PR-05",
            Severity::Warn,
            "one PR should close one primary issue",
        ));
    }

    // PR-10 plain-text Part of / Related links — INFO
    let text_links: Vec<String> = text_links_re()
        .captures_iter(body)
        .map(|c| c.get(1).unwrap().as_str().to_string())
        .collect();
    if !text_links.is_empty() {
        findings.push(Finding::new(
            "PR-10",
            Severity::Info,
            &format!(
                "Part of/Related #({}) 是纯文本，不产生 GitHub 关联；epic 关联通过 Fixes 的 sub-issue 层级或 UI development 面板",
                text_links.join(", ")
            ),
        ));
    } else {
        findings.push(Finding::new(
            "PR-10",
            Severity::Info,
            "no plain-text Part of/Related links",
        ));
    }

    // PR-06 type label + keyword suggestions (appears twice in Python — once
    // before P-11 and once before P-31; both emit identical findings.  We
    // replicate the first occurrence here.)
    let type_labels = cfg_type_labels(cfg);
    let label_set: HashSet<&str> = labels.iter().copied().collect();
    if type_labels.iter().any(|l| label_set.contains(l.as_str())) {
        findings.push(Finding::new("PR-06", Severity::Info, "type label present"));
    } else {
        findings.push(Finding::new(
            "PR-06",
            Severity::Fail,
            "no type label (expected one of the type set)",
        ));
    }
    let kw_map = cfg_keyword_suggestions(cfg);
    {
        let haystack = format!("{title}\n{body}").to_lowercase();
        let mut missing: Vec<String> = Vec::new();
        for (keyword, suggested) in &kw_map {
            if haystack.contains(&keyword.to_lowercase()) && !label_set.contains(suggested.as_str())
            {
                missing.push(suggested.clone());
            }
        }
        if !missing.is_empty() {
            // dedupe + sort (Python: sorted(set(missing)))
            let deduped: Vec<String> = {
                let s: HashSet<&str> = missing.iter().map(|s| s.as_str()).collect();
                let mut v: Vec<String> = s.into_iter().map(String::from).collect();
                v.sort();
                v
            };
            findings.push(Finding::new(
                "PR-06",
                Severity::Warn,
                &format!(
                    "based on content keywords, consider also labeling: {}",
                    deduped.join(" ")
                ),
            ));
        } else {
            findings.push(Finding::new(
                "PR-06",
                Severity::Info,
                "content keywords align with assigned labels",
            ));
        }
    }

    // PR-08 branch name — strip fork "user:" prefix before prefix check.
    // This is the fix for the real bug: a forkPR's head_ref is "user:branch",
    // and we must check the *branch* part, not "user:branch".
    let allowed = cfg_branch_prefixes(cfg);
    let branch = if head_ref.contains(':') {
        // Python: head_ref.rsplit(":", 1)[-1] — last segment after the final ":"
        head_ref.rsplit(':').next().unwrap_or(head_ref)
    } else {
        head_ref
    };
    if branch.is_empty() || !allowed.iter().any(|p| branch.starts_with(p.as_str())) {
        findings.push(Finding::new(
            "PR-08",
            Severity::Fail,
            &format!(
                "branch name not allowed: {branch} (allowed prefixes: {:?})",
                allowed
            ),
        ));
    } else {
        findings.push(Finding::new(
            "PR-08",
            Severity::Info,
            &format!("branch name OK: {branch} (prefixes: {:?})", allowed),
        ));
    }

    // PR-06 duplicate block (Python emits it a second time before P-31).
    // We replicate faithfully.
    if type_labels.iter().any(|l| label_set.contains(l.as_str())) {
        findings.push(Finding::new("PR-06", Severity::Info, "type label present"));
    } else {
        findings.push(Finding::new(
            "PR-06",
            Severity::Fail,
            "no type label (expected one of the type set)",
        ));
    }
    {
        let haystack = format!("{title}\n{body}").to_lowercase();
        let mut missing: Vec<String> = Vec::new();
        for (keyword, suggested) in &kw_map {
            if haystack.contains(&keyword.to_lowercase()) && !label_set.contains(suggested.as_str())
            {
                missing.push(suggested.clone());
            }
        }
        if !missing.is_empty() {
            let deduped: Vec<String> = {
                let s: HashSet<&str> = missing.iter().map(|s| s.as_str()).collect();
                let mut v: Vec<String> = s.into_iter().map(String::from).collect();
                v.sort();
                v
            };
            findings.push(Finding::new(
                "PR-06",
                Severity::Warn,
                &format!(
                    "based on content keywords, consider also labeling: {}",
                    deduped.join(" ")
                ),
            ));
        } else {
            findings.push(Finding::new(
                "PR-06",
                Severity::Info,
                "content keywords align with assigned labels",
            ));
        }
    }
    // Apply user-provided severity overrides (e.g. PR-05 demoted to WARN).
    crate::shared::apply_severity_overrides(&mut findings, cfg);

    findings
}

// ===========================================================================
// Tests
// ===========================================================================
